xbuild/src/devices/adb.rs

```diff
@@ -129,16 +129,25 @@ impl Adb {
     }
 
     /// To run a native activity use "android.app.NativeActivity" as the activity name
-    fn start(&self, device: &str, package: &str, activity: &str) -> Result<()> {
-        let status = self
+    fn start(
+        &self,
+        device: &str,
+        package: &str,
+        activity: &str,
+        launch_params: &[String],
+    ) -> Result<()> {
+        let status = dbg!(self
             .shell(device, None)
             .arg("am")
             .arg("start")
             .arg("-a")
             .arg("android.intent.action.MAIN")
             .arg("-n")
             .arg(format!("{}/{}", package, activity))
-            .status()?;
+            // Quote arguments for `am` so that they don't already get parsed by `adb`.
+            // TODO: use shlex quoting helpers? https://docs.rs/shlex/1.3.0/shlex/
+            .args(launch_params.iter().map(|s| format!("\"{s}\""))))
```

Comment:
This must be fixed. I quickly tested some hacky command strings and it already falls apart.

Reply:
For example, we have `-- -e cmdLine "some-subcommand --foo"`. If I insert the quotes such that the `--help` command becomes "free-standing" with `-- -e cmdLine "some-subcommand\" --foo\""` or similar, because my wrapping in code here "inverts" the quoted selection (end-result is `"subcommand" --foo""`), `adb` splits our command in two and `am start` doesn't know what `--foo` is. `shlex` should be able to re-quote and/or escape this string such that it always gets seen as `"subcommand --foo"` by `am start`.

Reply:
Fixed with `shlex`. Note that the above borked example at least prevents `adb` or `am` from parsing our string, and instead gives the `some-subcommand\" --foo\"` to our `Activity` as expected (to be read via https://github.com/rust-mobile/android-intent/pull/4). `shlex` on the reading side will "split" that as a single `["some-subcommand --foo"]` argument which obviously isn't valid. That's why it shouldn't have been double-quoted :)